selfdrive/controls/controlsd.py

```diff
@@ -164,25 +165,36 @@ def state_publish(self, car_events):
       cp_send.carParams = self.CP
       self.pm.send('carParams', cp_send)
 
+    # publish new carOutput
+    co_send = messaging.new_message('carOutput')
+    co_send.valid = True
+    if self.last_actuators is not None:
+      co_send.carOutput.actuatorsOutput = self.last_actuators
+    self.pm.send('carOutput', co_send)
+
   def controls_update(self, CC: car.CarControl):
     """control update loop, driven by carControl"""
 
     # send car controls over can
     now_nanos = self.can_log_mono_time if REPLAY else int(time.monotonic() * 1e9)
-    actuators_output, can_sends = self.CI.apply(CC, now_nanos)
+    self.last_actuators, can_sends = self.CI.apply(CC, now_nanos)
     self.pm.send('sendcan', can_list_to_can_capnp(can_sends, msgtype='sendcan', valid=self.CS.canValid))
 
-    return actuators_output
+    self.CC_prev = CC
 
 
 class Controls:
   def __init__(self, CI=None):
     self.card = CarD(CI)
 
-    self.CP = self.card.CP
+    self.params = Params()
+
+    with car.CarParams.from_bytes(self.params.get("CarParams", block=True)) as msg:
+      # TODO: this shouldn't need to be a builder
+      self.CP = msg.as_builder()
```

Comment:
why is it a builder?

Reply:
pulled this from the card refactor where it's needed for something with process replay

Reply:
Ah, it was a builder before as well. It's implicitly required since latcontrol_torque starts at the carParams values and modifies them overtime from torqued. We can either do this:

`self.torque_params = CP.lateralTuning.torque.as_builder()`

Or just make a new torque builder starting at CP's values.